python/chemiscope/structures/_shapes.py

```diff
@@ -0,0 +1,77 @@
+import warnings
+
+SHAPE_PARAMS = {
+    "ellipsoid": {"required": ["semiaxes"], "optional": []},
+    "sphere": {"required": ["radius"], "optional": []},
+    "polyhedron": {"required": ["vertices"], "optional": ["simplices"]},
+}
+
+
+def _get_shape_params(shape_key, dictionary):
+    shape_dict = {"kind": shape_key}
+    for p in SHAPE_PARAMS[shape_key]["required"]:
+        if p not in dictionary:
+            raise KeyError("Missing required parameter {} for {}.".format(p, shape_key))
+        else:
+            shape_dict[p] = dictionary[p]
+
+    for p in SHAPE_PARAMS[shape_key]["optional"]:
+        if p not in dictionary:
+            warnings.warn("Missing optional parameter {} for {}.".format(p, shape_key))
+        else:
+            shape_dict[p] = dictionary[p]
+    return shape_dict
+
+
+def _get_shape_params_i(shape_key, dictionary, i):
+    shape_dict = {"kind": shape_key}
+    for p in SHAPE_PARAMS[shape_key]["required"]:
+        if p not in dictionary:
+            raise KeyError("Missing required parameter {} for {}.".format(p, shape_key))
+        else:
+            shape_dict[p] = dictionary[p][i]
+
+    for p in SHAPE_PARAMS[shape_key]["optional"]:
+        if p not in dictionary:
+            warnings.warn("Missing optional parameter {} for {}.".format(p, shape_key))
+        else:
+            shape_dict[p] = dictionary[p][i]
+    return shape_dict
+
+
+def _decode_shape(frame):
```

Comment:
Is this a standard for shape storing in ASE, or something you made specifically for this? 

In general, we are moving away from storing everything in ASE Atoms and instead passing explicit arguments (see `extract_frame_properties` for example), so I would rather go for something like:

```py
explicit_shapes = [
    {"kind": "sphere", ...},
    {"kind": "ellispoid", ...},
]

chemiscope.write_input(..., shapes=explicit_shapes)

# if this makes sense
automatic_shapes = chemiscope.extract_shapes(frames)
chemiscope.write_input(..., shapes=automatic_shapes)

# could also be
gsd_shapes = chemiscope.shapes_from_gsd(gsd_frames)
chemiscope.write_input(..., shapes=gsd_shapes)
```

Reply:
In LAAMPS syntax, they will be stored in the ase-type items though. Can we keep as is?

Reply:
For which file format? LAMMPS data? Or some custom XYZ?

Reply:
LAMMPS data for non-spherical systems

Reply:
In this instance, what size/shape would you consider explicit shapes?

Reply:
I'm not sure I understand your question here. 

What I'm suggesting is to **not** automatically extract shapes from frames, and instead have the user explicitly extract it if they want it. 

The user would write something like:

```py
frames = ...

properties = chemiscope.extract_properties(frames)
shapes = {
    "bodies": chemiscope.extract_shapes(frames)
}

chemiscope.write_input(frames, properties, shapes=shapes)
```

This would mean adding a `shape` parameter to write_input, and documenting the `extract_shapes` function as only working with ASE Atoms created from LAMMPS simulations (for now).

Reply:
My question is about the data structure for sending shapes to the chemiscope input. Because the shapes ``belong'' to the frames right now, we have the option to define a universal shape-per-frame or a shape-per-particle, and so I feel it could get complicated with disecting this.

How tied are you to this API at this stage?

Reply:
right, I did not fully realized the shapes where inside the structures, this complicates things 😞. 

I'd still like to move away from automatically extracting shapes (since there is no unique convention for e.g. quaternions), and instead have the user explicitly say what they are trying to do. This way, we can easily add more shape extraction functions in the future.

The best I can think of right now would be this, which is quite verbose unfortunately:

```py
frames = ... 

structures = []
for frame in frames:
    # manually convert the structure
    structure = chemiscope.convert_structure(frame)
    # and manually add shapes to it
    structure["shapes"] = { 
        "bodies": chemiscope.shapes_from_ase_lammps(frame),
    } 
    structures.append(structure)

properties = chemiscope.extract_properties(frames)

chemiscope.write_input(structures, properties)
```

The good point of this API, is that we can add shapes corresponding to vector/tensorial properties quite easily with something like:

```py
for frame in frames:
    structure = chemiscope.convert_structure(frame)
    structure["shapes"] = { 
        "polarizability": [ellipsoid_from_tensor(p) for p in frame.polarizability],
        "dipoles": [arrow_from_vector(d) for d in frame.dipoles]
    } 
    structures.append(structure)
```


I'll think a bit more about how we could make the API nicer, but I'm pretty committed on making the convention used for shapes more explicit. I'm also willing to implement this myself once we agree on the API!

---

Also, if you have a very precise workflow, you can also create a `chemiscope_from_cersonsky_lab_workflow(signac_directory)` that would wrap the raw chemiscope API in something nicer!